File: leetcode/chain/203.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def removeElements(self, head: ListNode, val: int) -> ListNode:
        '''移除链表元素'''
        ret = ListNode()
        last = ret
        while head is not None:
            if head.val == val:  #满足删除条件
                head = head.next
                last.next = head  #指向
            else:
                last.next = head
                last = last.next  #挪位
                head = head.next
        return ret.next

File: leetcode/chain/203.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* remove_elements(ListNode* node, int val) {
    ListNode* last = new ListNode(0, node);
    ListNode* ret = last;
    while (node != nullptr) {
      if (node->val == val) {
        node = node->next;
        last->next = node;
      } else {
        last = node;
        node = node->next;
      }
    }
    return ret->next;
  }
};

int main() {
  ListNode* node = new ListNode(4);
  node->append(new ListNode(4));
  node->append(new ListNode(4));
  node->append(new ListNode(3));
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.remove_elements(node, 4);
  ret->print();
}